Create Staticd

donaldsharp opened the pull request:
Remove the handling of static routes from zebra into it's own daemon

donaldsharp commented:
I pushed this up so that I could easily get an idea of what would fail on redhat and to get an idea of what might fail in our topotests.

I know I need to write some documentation and work on some basic static route testing.

pguibert6WIND commented:
Hello,
yes some configuration will be welcome to understand why a static daemon.
static, you mean "ip/ipv6 route <blabla>" will no more be in zebra ?

mwinter-osr commented:
@donaldsharp ANVL (Protocol Compliance) Tests are updated to support staticd

rwestphal commented on staticd/static_routes.c:
```diff
@@ -0,0 +1,506 @@
+		static_zebra_nht_register(si, true);
+
+	si = rn->info;
+	static_zebra_route_add(rn, si->vrf_id, true);
```
On my tests staticd segfaults in this line (`si` is `NULL`) if I do this:
```
debian(config)# ip route 50.0.0.0/8 10.0.1.2
debian(config)# ip route 60.0.0.0/8 50.0.0.1
debian(config)#
root@debian:~#
root@debian:~# ip link add vrf-red type vrf table 10
root@debian:~# ip link set dev vrf-red up
root@debian:~# ip link set dev rt1-eth0 master vrf-red
```
Could you check what's the problem here?

donaldsharp replied:
fixed

donaldsharp commented on staticd/static_nht.c:
```diff
@@ -0,0 +1,78 @@
+	if (!stable)
+		return;
+
+	zlog_debug("Nexthop num: %d", nh_num);
```
fixed.

donaldsharp commented on staticd/static_nht.c:
```diff
@@ -0,0 +1,78 @@
+				si->nh_valid = !!nh_num;
+
+			if (orig != si->nh_valid) {
+				zlog_debug("Install the matching si");
```
fixed

donaldsharp commented on staticd/static_zebra.c:
```diff
@@ -0,0 +1,356 @@
+	case STATIC_IPV4_GATEWAY_IFNAME:
+	case STATIC_IFNAME:
+	case STATIC_BLACKHOLE:
+	case STATIC_IPV6_GATEWAY_IFNAME:
```
I disagree, we were not registering NEXTHOP_TYPE_IPV4_IFNAME and NEXTHOP_TYPE_IPV6_IFNAME before.  They can have no recursion since we are fully specifying the gateway and ifindex.

donaldsharp commented on staticd/static_zebra.c:
```diff
@@ -0,0 +1,356 @@
+	memset(&api, 0, sizeof(api));
+	api.vrf_id = vrf_id;
+	api.type = ZEBRA_ROUTE_STATIC;
+	api.safi = SAFI_UNICAST;
```
yep fixed.

donaldsharp commented on staticd/static_main.c:
```diff
@@ -0,0 +1,146 @@
+bool mpls_enabled;
+
+zebra_capabilities_t _caps_p[] = {
+	ZCAP_NET_RAW, ZCAP_BIND, ZCAP_NET_ADMIN,
```
I believe you are right.  removed.

donaldsharp commented on staticd/static_zebra.c:
```diff
@@ -0,0 +1,356 @@
+	struct zapi_route nhr;
+	afi_t afi = AFI_IP;
+
+	zapi_nexthop_update_decode(zclient->ibuf, &nhr);
```
fixed

donaldsharp commented on staticd/static_zebra.c:
```diff
@@ -0,0 +1,356 @@
+		/*
+		 * If we create a ecmp static route the
+		 * last distance and tag entered wins.  Why because
+		 * this cli choosen sucks
```
This is a long standing problem.  which is why I did not tackle this issue.  I've changed nothing from this perspective.  I think the solution is to add in nexthop_groups support for routes.

donaldsharp commented on staticd/static_zebra.c:
```diff
@@ -0,0 +1,356 @@
+			api_nh->gate = si->addr;
+			break;
+		case STATIC_IPV4_GATEWAY_IFNAME:
+			if (si->ifindex == IFINDEX_INTERNAL)
```
I disagree, NEXTHOP_TYPE_IPV4_IFINDEX cannot be not valid.  They are by definition valid.

donaldsharp commented on staticd/static_zebra.c:
```diff
@@ -0,0 +1,356 @@
+	uint32_t table_id;
+	char buf[PREFIX_STRLEN];
+
+	prefix2str(&p, buf, sizeof(buf));
```
correct, this was the start of some code I needed to finish up.  Finished up.

donaldsharp commented:
@pguibert6WIND I will add some documentation to this PR.  I thought I had done so.  apparently not.

qlyoung commented on staticd/static_main.c:
```diff
@@ -0,0 +1,151 @@
+ * Copyright (C) 2018 Cumulus Networks, Inc.
+ *               Donald Sharp
+ *
+ * FRR is free software; you can redistribute it and/or modify it
```
Please use the header given in the dev docs.
http://frrouting.readthedocs.io/projects/dev-guide/en/latest/workflow.html#source-file-header

donaldsharp replied:
Fixed.

qlyoung commented on staticd/static_memory.c:
```diff
@@ -0,0 +1,28 @@
+ * Copyright (C) 2018 Cumulus Networks, Inc.
+ *               Donald Sharp
+ *
+ * FRR is free software; you can redistribute it and/or modify it
```
license header

donaldsharp replied:
fixed

qlyoung commented on staticd/static_nht.c:
```diff
@@ -0,0 +1,75 @@
+ * Copyright (C) 2018 Cumulus Networks, Inc.
+ *               Donald Sharp
+ *
+ * FRR is free software; you can redistribute it and/or modify it
```
License header

donaldsharp replied:
fixed all headers

qlyoung commented on staticd/static_routes.h:
```diff
@@ -1,16 +1,14 @@
  * Free Software Foundation; either version 2, or (at your option) any
  * later version.
  *
- * Quagga is distributed in the hope that it will be useful, but
```
Since we're changing this, wanna replace the whole header?

donaldsharp replied:
I just removed the files, instead

qlyoung commented on staticd/static_vrf.c:
```diff
@@ -0,0 +1,185 @@
+ * Copyright (C) 2018 Cumulus Networks, Inc.
+ *               Donald Sharp
+ *
+ * FRR is free software; you can redistribute it and/or modify it
```
license header

donaldsharp replied:
fixed.

qlyoung commented on staticd/static_vty.c:
```diff
@@ -0,0 +1,1251 @@
+						"%% Malformed source address\n");
+				else
+					zlog_warn(
+						"%s: Malformed Source address: %s",
```
`Source` vs `source`

donaldsharp replied:
fixed

pguibert6WIND commented:
There will be a change to think of. static daemon will have to be aware on the vrf backend type.
so that some commands may / may not be available for the daemon

donaldsharp commented:
I've already started the work for this communication via the ZEBRA_CAPABILITIES message.  Feel free to start adding to it.

donaldsharp commented:
I've updated to latest and brought in the changes in zebra that were moved.
I've added documentation
I believe that renato's concern about being able to handle zebra post connection just works given the way staticd holds routes it cannot insert due to insufficient information.

Have I missed anything?

louberger commented:
@donaldsharp you still have one code conventions error:
===============================================
ERROR: space prohibited before that close parenthesis ')'
#111: FILE: /tmp/f1-25096/static_main.c:111:
+		.privs = &static_privs, )

donaldsharp commented:
@louberger when I remove that space the script generates a different error.  Known issue with it there.  I'm consistent with other places in the code for that structure

louberger commented:
any chance you can fix the script as part of the PR?  If not I guess 
need to create a PR and assign to ql...


On 7/29/2018 1:09 PM, Donald Sharp wrote:
>
> @louberger <https://github.com/louberger> when I remove that space the 
> script generates a different error. Known issue with it there. I'm 
> consistent with other places in the code for that structure
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/FRRouting/frr/pull/2122#issuecomment-408691495>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AGRWyAuiVBey0UqHBnq3iSN1eerf5hO9ks5uLewsgaJpZM4TkXPH>.
>

eqvinox commented on staticd/static_vrf.c:
```diff
@@ -0,0 +1,196 @@
+
+static int static_vrf_disable(struct vrf *vrf)
+{
+	return 0;
```
What happens when a VRF is disabled and the ID is reused for another VRF? We're not clearing out the VRF ID that was set in `static_vrf_enable()`, is it possible to get conflicts here or is it OK to keep the IDs cached?

(Also, the answer to this question should be a code comment here because it'll be hard to understand when someone else looks at this in a year...)

donaldsharp replied:
well it will never be reused( at least most likely will never be reused as that it's a uint32_t number that is increased monotomically per new interface ), I think your point stands though.

What the vrf code does, as I recall, if we shutdown the vrf we disable routes and when it is turned back on we scan the code base and fixup the nhop vrf id's.( at least that is what I recall programming for static routes 4-5 months back ).  All daemons that are vrf aware are probably going to have to be aware of this though.

donaldsharp commented:
@louberger I'm not currently interested in fixing that script.  We can file an issue and get it addressed though.  I'm not sure what the answer is because I think it's valid code either way.

louberger commented:
If it's valid either way the script needs to be fixed to allow for it. Just 
open an issue, note it in this PR, and let's move on. I'm not the style 
police, but we do have a tools based process that needs to work in order 
for it to be of value.


----------
On July 30, 2018 8:21:45 AM Donald Sharp <notifications@github.com> wrote:

> @louberger I'm not currently interested in fixing that script.  We can file 
> an issue and get it addressed though.  I'm not sure what the answer is 
> because I think it's valid code either way.
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/FRRouting/frr/pull/2122#issuecomment-408844513

eqvinox commented:
The fix (and correct thing to do here) is to make it
```
FRR_DAEMON_INFO(staticd, STATIC, .vty_port = STATIC_VTY_PORT,
[...]
        .privs = &static_privs,
)
```

The rationale for this is, both for the `,` and the `)` on a separate line, to have one line each be a functional unit. This is to say that when a patch wants to insert another element, it just adds 1 line before the `)`.  But this only works if you neither have to change the previous line to add the `,` nor have to add the extra linebreak before the `)`.

In general, writing lists and struct initializers with a trailing comma and a newline before the closing `)` or `}` is almost always good style.

Other than that: LGTM, please add the linebreak and merge ASAP :)

eqvinox commented:
Actually went and pushed the fix on your repo (you had "allow maintainers to push changes" ticked :D), let's see if the script still spits an error :)
